drivers/counter/counter_native_posix.c

```diff
@@ -107,17 +171,27 @@ static int ctr_set_alarm(const struct device *dev, uint8_t chan_id,
 		return -ENOTSUP;
 	}
 
-	pending_alarm = *alarm_cfg;
-	is_alarm_pending = true;
+	if (alarm_cfg->ticks > top_cfg.ticks) {
+		posix_print_warning(
+			"channel %u alarm value exceeds top value (%d)\n",
+			chan_id, top_cfg.ticks);
+		return -EINVAL;
+	}
 
-	if (!(alarm_cfg->flags & COUNTER_ALARM_CFG_ABSOLUTE)) {
-		pending_alarm.ticks =
-			hw_counter_get_value() + pending_alarm.ticks;
+	if (alarms[chan_id].active) {
+		posix_print_warning("channel %u is busy\n", chan_id);
+		return -EBUSY;
 	}
 
-	hw_counter_set_target(pending_alarm.ticks);
-	irq_enable(COUNTER_EVENT_IRQ);
+	alarms[chan_id].config = *alarm_cfg;
+	alarms[chan_id].active = true;
+
+	if (!(alarm_cfg->flags & COUNTER_ALARM_CFG_ABSOLUTE)) {
+		alarms[chan_id].config.ticks =
```

Comment:
should this be done before the top check? (or include a top check after offsetting?)

Reply:
Should I also check a possible alarm ticks overflow or user of this driver should worry about that?

Reply:
Given that this is a driver API question, @nordic-krch ^^ please see above as counter driver maintainer

Reply:
you should wrap it here.

Reply:
Yes, it should. In case of exceeding the top value by adding the offset, the alarm value will be trimmed to conform counter range.